tensorflow/contrib/go/graph.go

```diff
@@ -0,0 +1,605 @@
+package tensorflow
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/golang/protobuf/proto"
+
+	pb "github.com/tensorflow/tensorflow/tensorflow/contrib/go/proto"
+)
+
+const (
+	cOpsProtobufDefsPath = "/usr/local/tensorlow/ops.pbtxt"
+)
+
+// Graph Representation of the computation graph.
+type Graph struct {
+	def *pb.GraphDef
+
+	availableOps map[string]*pb.OpDef
+	constants    map[string]*Tensor
+	variables    map[string]*Tensor
+}
+
+// GraphNode Representation of one of the nodes of the TensorFlow Graph.
+// A node takes zero or more Tensors, performs some computation, and
+// produces zero or more Tensors.
+type GraphNode struct {
+	ref          *pb.NodeDef
+	def          *pb.NodeDef
+	outDataTypes map[string]DataType
+}
+
+// NewGraph Returns an initialized instance of the Graph struct.
+func NewGraph() *Graph {
+	return &Graph{
+		def:          new(pb.GraphDef),
+		availableOps: make(map[string]*pb.OpDef),
+		constants:    make(map[string]*Tensor),
+		variables:    make(map[string]*Tensor),
+	}
+}
+
+// NewGraphFromText Returns a new graph populated with the deserialization of
+// the provided graph string.
+func NewGraphFromText(graphStr string) (gr *Graph, err error) {
+	gr = NewGraph()
+	err = proto.UnmarshalText(graphStr, gr.def)
+
+	return
+}
+
+// LoadGraphFromFile Loads a Graph from the file on the specified path.
+func LoadGraphFromFile(path string) (gr *Graph, err error) {
+	graphStr, err := ioutil.ReadFile(path)
+	if err != nil {
+		return
+	}
+
+	gr = NewGraph()
+	err = proto.Unmarshal(graphStr, gr.def)
+
+	return
+}
+
+// LoadGraphFromTextFile Loads a Graph as plain text from the file on the specified
+// path.
+func LoadGraphFromTextFile(path string) (gr *Graph, err error) {
+	graphStr, err := ioutil.ReadFile(path)
+	if err != nil {
+		return
+	}
+
+	return NewGraphFromText(string(graphStr))
+}
+
+// Op Adds a new Node to the Graph with the specified operation, this function
+// could return an error if any of the mandatory attributes is not be present
+// or the value is not the expected for this attribute.
+func (gr *Graph) Op(opName string, name string, input []*GraphNode, device string, attrs map[string]interface{}) (node *GraphNode, err error) {
```

Comment:
I agree with you, I'm going to use explicit return values for this method. This is more a personal preference in Go, if you are changing the function definition really often, the params that you are returning, etc, assign and return makes more easy to work with the code so you don't need to change all the returned values every time the returned params changes, but since here we are returning mostly errors generated inside this function I think that is better if the error and the return goes together.

I also changed the assignations and conditionals in the same line, the indentation was super ugly, I don't know what I was thinking when I did that ¯\_(ツ)_/¯

Perhaps I'm going to split this in two, the logic for the operation and the logic for the attribute values casting.